#1320 Add module TypeScript definitions

samueldmeyer opened the pull request:
Fixes: #1320 
 
Developer's Certificate of Origin 1.1

       By making a contribution to this project, I certify that:

       (a) The contribution was created in whole or in part by me and I
           have the right to submit it under the Apache License 2.0; or

       (b) The contribution is based upon previous work that, to the best
           of my knowledge, is covered under an appropriate open source
           license and I have the right under that license to submit that
           work with modifications, whether created in whole or in part
           by me, under the same open source license (unless I am
           permitted to submit under a different license), as indicated
           in the file; or

       (c) The contribution was provided directly to me by some other
           person who certified (a), (b) or (c) and I have not modified
           it.

       (d) I understand and agree that this project and the contribution
           are public and that a record of the contribution (including all
           personal information I submit with it, including my sign-off) is
           maintained indefinitely and may be redistributed consistent with
           this project or the open source license(s) involved.

srikant-ch5 commented:
Hi @samueldmeyer Thanks for the PR.

Could you please fix the DCO issue and also attach the issue [1320](https://github.com/elyra-ai/canvas/issues/1320) like below in the PR above Dev Certificate ?

<img width="318" alt="Screenshot 2025-01-08 at 11 21 55 AM" src="https://github.com/user-attachments/assets/5b68e7a4-fd0e-4ce1-ba3b-8a7b55a36d04" />

samueldmeyer commented:
@srikant-ch5 Is there anything else I need to do for this pull request?

tomlyn commented:
Hi @samueldmeyer  -- I think I was supposed to look through this. Apologies for not doing so, I've been concentrating on some of the features you asked for. 

At first glance I wonder whether we need some tests (presumably Jest tests) to test at least some of this? 

Also, I think @matthoward366, @nmgokhale and @caritaou should review the Common Properties parts of this since they will need to maintain this Type script API going forward.

samueldmeyer commented:
@tomlyn I added tests using tsd, which is the standard for TypeScript definitions inside the library. I didn't test everything because a lot of that would simply be a repeat of what is in the definitions, but I did test the parts that are more complicated and generally made it extensible for adding tests. I added the tests to `npm test`, but you could tell me another way you want it run.

srikant-ch5 commented:
@samueldmeyer I have updated test-harness to use Typescript Definitions by updating below in test harness `package.json` to use `index.d.ts` from `common-canvas`

<img width="561" alt="Screenshot 2025-02-06 at 8 29 18 PM" src="https://github.com/user-attachments/assets/8aaa0fd3-9e9a-4879-b7aa-5f83d608ca45" />

After making few more changes to the setup I can see that Common Canvas and Common Properties are working fine.


https://github.com/user-attachments/assets/11fd5f1f-5397-42a9-a5d1-39c79106b60e

If there are any other ways to test this PR, please feel free to let me know so that I can test the functionality.

samueldmeyer commented:
@srikant-ch5 The only way to test TypeScript definitions is to use it with TypeScript code. For that purpose, we have a test file: `index.test-d.ts`. You can tell me if there are any additional tests we should add to verify methods have correct types in `canvas_modules/common-canvas/types/common-properties-controller.ts`

tomlyn commented:
HI @srikant-ch5 -- I am making some changes to this PR. I recommend you pause your work on it for a while, at least until I have pushed my changes. If you have time you could start to think about what tests we could add to `index.test-d.ts` as suggested by @samueldmeyer

samueldmeyer commented on canvas_modules/common-canvas/types/common-canvas.ts:
```diff
@@ -366,14 +363,17 @@ export declare class CanvasController {
    */
   setPipelineFlowPalette(
-    palette: HttpsApiDataplatformIbmComSchemasCommonCanvasPalettePaletteV3SchemaJson
+    palette: RelaxedPipelineFlowPalette
```
Can you make this `RelaxedPipelineFlowPalette | PipelineFlowPalette` just to guide those of us who generate this programmatically?

samueldmeyer commented on canvas_modules/common-canvas/types/common-canvas.ts:
```diff
@@ -16,43 +16,80 @@
+  CanvasNode,
+  CanvasSupernode,
+  CanvasComment,
+  CanvasLink,
```
Can you export `CanvasCommentLink`, `CanvasNodeLink`, and `CanvasAssociationLink` as well?

samueldmeyer commented on canvas_modules/common-canvas/locales/command-actions/locales/index.d.ts:
```diff
@@ -0,0 +1,18 @@
+ * limitations under the License.
+ */
+
+declare const messages: Record<string, string>;
```
On testing, I realize this should be `Record<string, Record<string, string>>;`

tomlyn replied:
Done

samueldmeyer commented on canvas_modules/common-canvas/locales/common-canvas/locales/index.d.ts:
```diff
@@ -0,0 +1,18 @@
+ * limitations under the License.
+ */
+
+declare const messages: Record<string, string>;
```
On testing, I realize this should be `Record<string, Record<string, string>>;`

tomlyn replied:
Done

samueldmeyer commented on canvas_modules/common-canvas/locales/common-properties/locales/index.d.ts:
```diff
@@ -0,0 +1,18 @@
+ * limitations under the License.
+ */
+
+declare const messages: Record<string, string>;
```
On testing, I realize this should be `Record<string, Record<string, string>>;`

samueldmeyer commented on canvas_modules/common-canvas/locales/notification-panel/locales/index.d.ts:
```diff
@@ -0,0 +1,18 @@
+ * limitations under the License.
+ */
+
+declare const messages: Record<string, string>;
```
On testing, I realize this should be `Record<string, Record<string, string>>;`

tomlyn replied:
Done

samueldmeyer commented on canvas_modules/common-canvas/locales/toolbar/locales/index.d.ts:
```diff
@@ -0,0 +1,18 @@
+ * limitations under the License.
+ */
+
+declare const messages: Record<string, string>;
```
On testing, I realize this should be `Record<string, Record<string, string>>;`

tomlyn replied:
Done

samueldmeyer commented on canvas_modules/common-canvas/locales/palette/locales/index.d.ts:
```diff
@@ -0,0 +1,18 @@
+ * limitations under the License.
+ */
+
+declare const messages: Record<string, string>;
```
On testing, I realize this should be `Record<string, Record<string, string>>;`

tomlyn replied:
Done

samueldmeyer commented on canvas_modules/common-canvas/types/common-canvas.ts:
```diff
@@ -0,0 +1,2872 @@
+  selectedObjects: Record<string, unknown>[];
+  /** @deprecated */
+  selectedObjectIds: string[];
+  [key: string]: unknown;
```
Based on `SVGCanvasRenderer.openContextMenu`, it seems like `type`, `targetObject`, `pipelineId` and a few others are included if the `editSource` is `contextmenu`. Is that a level of detail we should include? Or should we skip that for now? I only bring this up because you used `targetObject` in `editActionHandler` in `canvas_modules/harness/src/client/components/custom-canvases/prompt/prompt-canvas.jsx`

tomlyn replied:
This is a tough one because the properties of the `EditActionData` object can vary quite a lot based on:
* Is it the parameter for `beforeEditActionHandler` or `editActionHandler`
* Which edit action is being performed.
* What originated the action.
I'll open a new issue to do a better job of specifying these to add more detail.

samueldmeyer replied:
Yes, let's leave that as a future enhancement instead of having it block this PR.

tomlyn replied:
I opened this issue to address this in the future: https://github.ibm.com/NGP-TWC/wdp-abstract-canvas/issues/3979

samueldmeyer commented on canvas_modules/common-canvas/types/common-canvas.ts:
```diff
@@ -1143,138 +1395,220 @@ export declare class CanvasController {
+      linkName?: string;
     },
     pipelineId?: string
   ): CanvasLink[];
```
Here the correct thing tool is a type overload so that the output type depends on the inputs. You end up defining it twice, one for each case.

Also, this seems to be another case where the schema has snake_case but the code has camelCase for nodeLink.

```
createNodeLinks(
    data: {
      id?: string;
      type: "nodeLink"
 nodes: {
        id?: string;
        portId?: string;
        srcPos?: {
          x_pos: number;
          y_pos: number;
        };
      }[];
      targetNodes: {
        id?: string;
        portId?: string;
        trgPos?: {
          x_pos: number;
          y_pos: number;
        };
      }[],
      class_name?: string;
      linkName?: string;
    },
    pipelineId?: string
  ): CanvasNodeLink[];
createNodeLinks(
    data: {
      id?: string;
      type: "associationLink"
 nodes: {
        id?: string;
        portId?: string;
        srcPos?: {
          x_pos: number;
          y_pos: number;
        };
      }[];
      targetNodes: {
        id?: string;
        portId?: string;
        trgPos?: {
          x_pos: number;
          y_pos: number;
        };
      }[],
      class_name?: string;
      linkName?: string;
    },
    pipelineId?: string
  ): CanvasAssociationLink[];
```

samueldmeyer commented on canvas_modules/common-canvas/types/common-canvas-schema-types/canvas-info-v3.ts:
```diff
@@ -0,0 +1,1205 @@
+/**
+ * Node link definition
+ */
+export interface CanvasNodeLink {
```
If they are private, I'll update my code to look them up based on the srcNodeId and trgNodeId

tomlyn replied:
They don't have to be private. But, like the height and width mentioned above, they are added during internal processing rather than being specified through the API.

samueldmeyer replied:
Then this is another good case for `readonly`, even if it requires manual intervention.

tomlyn commented on canvas_modules/common-canvas/types/common-canvas-schema-types/canvas-info-v3.ts:
```diff
@@ -0,0 +1,1205 @@
+  /**
+   * x-position of this canvas execution node.
+   */
+  x_pos?: number;
+  /**
+   * y-position of this canvas execution node.
+   */
+  y_pos?: number;
```
OK - I added a comment to the schema file which should get generated into the canvas-info.ts file.

tomlyn commented on canvas_modules/common-canvas/types/common-canvas-schema-types/canvas-info-v3.ts:
```diff
@@ -0,0 +1,1205 @@
+ */
+/* eslint-disable */
+
+export type CanvasNode =
```
I added it as optional and added `"readOnly": true` to the properties in the schema but they don't seem to be converted to readonly in TS.

samueldmeyer replied:
Looking at the library issues, the lack of `readonly` seems to be a limitation of the library (and possibly the JSON schema spec, depending on who you ask). The best we could do is to add a type that depends on CanvasNode but manually adds them as readonly properties.

tomlyn commented on canvas_modules/common-canvas/types/pipeline-flow-v3.ts:
```diff
@@ -0,0 +1,512 @@
+/**
+ * WDP Pipeline Flow Schema
+ */
+export interface HttpsApiDataplatformIbmComSchemasCommonPipelinePipelineFlowPipelineFlowV3SchemaJson {
```
Done

tomlyn commented on canvas_modules/common-canvas/types/index.d.ts:
```diff
@@ -0,0 +1,2810 @@
+interface ShapeDecoration {
+  id: string;
+  path?: string;
+  position?: NodePosition | LinkPosition;
```
Done

tomlyn commented on canvas_modules/common-canvas/types/index.d.ts:
```diff
@@ -0,0 +1,2810 @@
+  label_single_line?: boolean;
+  label_max_characters?: number;
+  label_allow_return_key?: boolean;
+  position?: NodePosition | LinkPosition;
```
Done

tomlyn commented on canvas_modules/common-canvas/types/index.d.ts:
```diff
@@ -0,0 +1,2810 @@
+interface JsxDecoration {
+  id: string;
+  jsx: ReactNode;
+  position: NodePosition | LinkPosition;
```
Done

tomlyn commented on canvas_modules/common-canvas/types/index.d.ts:
```diff
@@ -0,0 +1,2810 @@
+  | ImageDecoration
+  | LabelDecoration
+  | ShapeDecoration
+  | JsxDecoration;
```
Done

tomlyn commented on canvas_modules/common-canvas/types/index.d.ts:
```diff
@@ -0,0 +1,2810 @@
+    app_data: T;
+    height: number;
+    width: number;
+  };
```
Done

tomlyn commented on canvas_modules/common-canvas/types/index.d.ts:
```diff
@@ -0,0 +1,2810 @@
+   * @param url
+   */
+  getExternalPipelineFlow(url: string): Record<string, unknown>;
+  /**
```
Done

tomlyn commented on canvas_modules/common-canvas/types/index.d.ts:
```diff
@@ -0,0 +1,2810 @@
+  /**
+   * @return the pipeline object for the pipeline Id passed in.
+   */
+  getPipeline(pipelineId: string): Record<string, unknown>;
```
Done

tomlyn commented on canvas_modules/common-canvas/types/index.d.ts:
```diff
@@ -0,0 +1,2810 @@
+  /**
+   * Returns the flow validation messages for the pipeline ID passed in.
+   */
+  getFlowMessages(pipelineId: string): Record<string, unknown>;
```
Done

tomlyn commented on canvas_modules/common-canvas/types/index.d.ts:
```diff
@@ -0,0 +1,2810 @@
+  setSelections(
+    newSelection: string[] | Record<string, unknown>[],
+    pipelineId?: string
+  ): void;
```
Done

tomlyn commented on canvas_modules/common-canvas/types/index.d.ts:
```diff
@@ -0,0 +1,2810 @@
+  deleteObjects(
+    objectIds: (string | CanvasNode | CanvasComment)[],
+    pipelineId: string
+  ): void;
```
Done

tomlyn commented on canvas_modules/common-canvas/types/common-canvas.ts:
```diff
@@ -1143,138 +1395,220 @@ export declare class CanvasController {
    */
   deleteLink(
-    link: string | CanvasLink | Record<string, unknown>,
+    link: CanvasLink,
```
Did you see my question here /\ ?

samueldmeyer replied:
Yes, I should just update my code. I have no idea why it worked either. Maybe it was just a side effect of some other code I wrote that deleted the link.

samueldmeyer commented on canvas_modules/common-canvas/types/common-canvas.ts:
```diff
@@ -530,26 +567,33 @@ export declare class CanvasController {
+   * @param pipelineId - Optional. The ID of the pipeline of the nodes.
+   *                     Defaults to the currently displayed pipeline.
+   */
+  deselectAll(pipelineId): void;
```
pipelineId needs a type of `string`

tomlyn replied:
Done

samueldmeyer reviewed (commented):
Only one change still needed! Otherwise it looks good

samueldmeyer commented:
One extra note: @elyra/pipeline-schemas needs the dependency version updated to include the types that are now there

tomlyn commented:
OK I updated the version in package.json and pushed my latest changes

samueldmeyer commented:
@tomlyn Did you mean to push another commit? I don't see a new one

samueldmeyer commented:
It all looks good now!

tomlyn reviewed (approved):
LGTM
